### PR TITLE
Elasticsearch consumer/sink

### PR DESCRIPTION
**Pull request by sobychacko**
* Elasticsearch consumer
 * Elasicsearch sink based on the consumer
 * Adding tests to verify

### CODE REVIEW

File: functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+					logger.error("Error occurred while indexing the document", e);
+				}
+			});
+		}
+		else {
+			try {
+				restHighLevelClient.index(request, RequestOptions.DEFAULT);
+			}
+			catch (IOException e) {
+				logger.error("Error occurred while indexing the document", e);
```
Comment (dturanski) [inline functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerConfiguration.java:105]: No exception thrown?

File: functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		if (isAsync) {
+			restHighLevelClient.indexAsync(request, RequestOptions.DEFAULT, new ActionListener<IndexResponse>() {
+				@Override
+				public void onResponse(IndexResponse indexResponse) {
+					logger.debug("Document with ID: " + indexResponse.getId() + " has been indexed");
+				}
+
+				@Override
+				public void onFailure(Exception e) {
+					logger.error("Error occurred while indexing the document", e);
```
Comment (dturanski) [inline functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerConfiguration.java:96]: No exception thrown?

File: functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerProperties.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	/**
+	 * Name of the index.
+	 */
+	String index;
+
+	/**
+	 * Indicates the shard to route to.
+	 * If not provided, this resolves to the ID used on the document.
+	 */
+	String routing;
```
Comment (dturanski) [inline functions/consumer/elasticsearch-consumer/src/main/java/org/springframework/cloud/fn/consumer/elasticsearch/ElasticsearchConsumerProperties.java:39]: These could be expressions also but we can defer that until someone asks for it

File: applications/sink/elasticsearch-sink/README.adoc [hunk 4]
```diff
@@ -0,0 +1,42 @@
+//tag::ref-doc[]
+= Elasticsearch Sink
+
+Sink that index documents into Elasticsearch.
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply applications/sink/elasticsearch-sink/README.adoc:4]: The operation itself is called indexing. If that sounds confusing from our perspective, I can change to something like `produce` or `publish`.

**Comment by dturanski** [comment]
merged via 221ff6d

**Pull request closed without merging**